enhancements/machine-config/user-data-secret-managed.md

```diff
@@ -0,0 +1,113 @@
+---
+title: Machines user-data managed by the MCO
+authors:
+  - "@runcom"
+reviewers:
+  - "@yuqi-zhang"
+  - "@cgwalters"
+  - "@enxebre"
+  - "@JoelSpeed"
+approvers:
+  - "@yuqi-zhang"
+  - "@cgwalters"
+  - "@enxebre"
+  - "@JoelSpeed"
+creation-date: 2020-06-09
+last-updated: yyyy-mm-dd
+status: provisional|implementable|implemented|deferred|rejected|withdrawn|replaced
+see-also:
+  - "/enhancements/this-other-neat-thing.md"  
+replaces:
+  - "/enhancements/that-less-than-great-idea.md"
+superseded-by:
+  - "/enhancements/our-past-effort.md"
+---
+
+# Machines user-data managed by the MCO
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+
+## Summary
+
+Currently the installer creates [pointer Ignition configs][1] which end up as [user][2] [data][3] secrets [referenced by the installer-generated Machine(Sets)][4].  That makes things like the Ignition v2 (spec 3) transition tricky.
+The key point is that the Secret holding the initial stub ignition is completely unmanaged after it's created by the installer.
+In order to upgrade to the new Ignition spec, the MCO needs a way to update the Secret to the new Ignition spec.
+
+
+[1]: https://github.com/openshift/installer/blob/093ca65398fe567bdf63322894496cebbe3d2ade/pkg/asset/ignition/machine/node.go#L30-L36
+[2]: https://github.com/openshift/installer/blob/093ca65398fe567bdf63322894496cebbe3d2ade/pkg/asset/machines/master.go#L161-L170
+[3]: https://github.com/openshift/installer/blob/093ca65398fe567bdf63322894496cebbe3d2ade/pkg/asset/machines/worker.go#L197-L205
+[4]: https://github.com/openshift/installer/blame/093ca65398fe567bdf63322894496cebbe3d2ade/docs/user/aws/install_upi.md#L231-L232
+
+## Motivation
+
+The goal for this enhancement is to have the Secret that contains the stub ignition configuration managed by the MCO. The MCO will later be able to update the stub to a newer version of the spec during an upgrade.
+
+### Goals
+
+The MCO owns the lifecycle of the Secret containing the pointer config. The installer is still responsible to create the initial version of the Secret as that is required to support UPI installation by generating the initial ignition configs until the installer provisions the control plane via MAO and not TF. The MCO takes ownership of the Secret containing the stub ignition once the cluster is up and running. When we'll update to the new Ignition version, the MCO can update the Secret.
```

Comment:
>as that is required to support UPI installation

Isn't this also true for IPI? i.e controlPlane instances are created from terraform in any scenario atm.

In terms of removing this from the installer, couldn't the mco use its render capabilities to generate these secrets?

Reply:
we will at some point but the MCO runs too late (bootkube) - this enhacement makes sure we touch few places with little changes for 4.6 and allows us to re-discuss about this post-4.6

Reply:
So if my understanding is correct, can we have a specific goal saying this is a stopgap solution to support upgrades from 4.5 to 4.6 because a new version of ign is being shipped within the OS?

Is there atm any other enhancement showing the long term plan for managing ign/bootimage upgrades?

Reply:
> So if my understanding is correct, can we have a specific goal saying this is a stopgap solution to support upgrades from 4.5 to 4.6 because a new version of ign is being shipped within the OS?

That is exactly the goal of this doc - I was torn if it needed an enhancement at all also. Yes, I can clarify this in this enhancement

> Is there atm any other enhancement showing the long term plan for managing ign/bootimage upgrades?

see https://github.com/openshift/enhancements/pull/201